File: pennyway-socket/src/main/java/kr/co/pennyway/socket/common/aop/PreAuthorizeAspect.java
package kr.co.pennyway.socket.common.aop;

import kr.co.pennyway.socket.common.annotation.PreAuthorize;
import kr.co.pennyway.socket.common.exception.PreAuthorizeErrorCode;
import kr.co.pennyway.socket.common.exception.PreAuthorizeErrorException;
import kr.co.pennyway.socket.common.util.PreAuthorizeSpELParser;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.reflect.MethodSignature;
import org.springframework.context.ApplicationContext;
import org.springframework.stereotype.Component;

import java.lang.reflect.Method;
import java.security.Principal;
import java.util.stream.Stream;

@Slf4j
@Aspect
@Component
@RequiredArgsConstructor
public class PreAuthorizeAspect {
    private final ApplicationContext applicationContext;

    /**
     * {@link PreAuthorize} 어노테이션이 붙은 메서드를 가로채고 인증/인가를 수행합니다.
     *
     * @param joinPoint 가로챈 메서드의 실행 지점
     * @return 인증/인가가 성공하면 원래 메서드의 실행 결과, 실패하면 UnauthorizedResponse
     * @throws Throwable 메서드 실행 중 발생한 예외
     */
    @Around("@annotation(kr.co.pennyway.socket.common.annotation.PreAuthorize)")
    public Object execute(final ProceedingJoinPoint joinPoint) throws Throwable {
        MethodSignature signature = (MethodSignature) joinPoint.getSignature();
        Method method = signature.getMethod();
        PreAuthorize preAuthorize = method.getAnnotation(PreAuthorize.class);

        Principal principal = extractPrincipal(joinPoint.getArgs());

        boolean isAuthorized = PreAuthorizeSpELParser.evaluate(preAuthorize.value(), method, joinPoint.getArgs(), applicationContext);

        if (!isAuthorized) {
            handleUnauthorized(principal, preAuthorize);
        }

        return joinPoint.proceed();
    }

    /**
     * 메서드 인자에서 Principal 객체를 추출합니다.
     *
     * @param args 메서드 인자 배열
     * @return 찾은 Principal 객체, 없으면 null
     */
    private Principal extractPrincipal(Object[] args) {
        return Stream.of(args)
                .filter(arg -> arg instanceof Principal)
                .map(arg -> (Principal) arg)
                .findFirst()
                .orElse(null);
    }

    /**
     * 인증/인가 실패 시 수행할 동작을 정의합니다.
     */
    private void handleUnauthorized(Principal principal, PreAuthorize preAuthorize) {
        if (preAuthorize.value().contains(PreAuthorizeSpELParser.SpELFunction.IS_AUTHENTICATED.getName())) {
            log.warn("인증 실패: {}", principal);
            throw new PreAuthorizeErrorException(PreAuthorizeErrorCode.UNAUTHENTICATED);
        }
    }
}

File: pennyway-socket/src/main/java/kr/co/pennyway/socket/common/aop/PreAuthorizeAspect.kt
package kr.co.pennyway.socket.common.aop;

import kr.co.pennyway.socket.common.annotation.PreAuthorize
import kr.co.pennyway.socket.common.exception.PreAuthorizeErrorCode
import kr.co.pennyway.socket.common.exception.PreAuthorizeErrorException
import kr.co.pennyway.socket.common.util.PreAuthorizeSpELParser
import kr.co.pennyway.socket.common.util.logger
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.reflect.MethodSignature
import org.springframework.context.ApplicationContext
import org.springframework.stereotype.Component
import java.lang.reflect.Method
import java.security.Principal

@Aspect
@Component
class PreAuthorizeAspect(private val applicationContext: ApplicationContext) {
    private val log = logger()

    /**
     * {@link PreAuthorize} 어노테이션이 붙은 메서드를 가로채고 인증/인가를 수행합니다.
     *
     * @param joinPoint 가로챈 메서드의 실행 지점
     * @return 인증/인가가 성공하면 원래 메서드의 실행 결과, 실패하면 UnauthorizedResponse
     * @throws Throwable 메서드 실행 중 발생한 예외
     */
    @Around("@annotation(kr.co.pennyway.socket.common.annotation.PreAuthorize)")
    fun execute(joinPoint: ProceedingJoinPoint): Any = with(joinPoint) {
        (signature as? MethodSignature)
            ?.method
            ?.let { method -> validateAccess(method, this) }
            ?: throw IllegalStateException("PreAuthorize는 메서드에만 적용할 수 있습니다")
    }

    private fun validateAccess(method: Method, joinPoint: ProceedingJoinPoint): Any {
        val preAuthorize = method.requireAnnotation<PreAuthorize>()
        val principal = joinPoint.args.findPrincipal()

        return evaluateAccess(
            principal = principal,
            preAuthorize = preAuthorize,
            method = method,
            args = joinPoint.args
        ).let { joinPoint.proceed() }
    }

    private fun evaluateAccess(
        principal: Principal?,
        preAuthorize: PreAuthorize,
        method: Method,
        args: Array<Any>
    ) = PreAuthorizeSpELParser
        .evaluate(preAuthorize.value, method, args, applicationContext)
        .also { result -> handleEvaluationResult(result, principal) }

    private fun handleEvaluationResult(
        result: PreAuthorizeSpELParser.EvaluationResult,
        principal: Principal?
    ) = when (result) {
        is PreAuthorizeSpELParser.EvaluationResult.Permitted -> Unit
        is PreAuthorizeSpELParser.EvaluationResult.Denied.Unauthenticated -> {
            log.warn("인증 실패: {}", principal)
            throw PreAuthorizeErrorException(PreAuthorizeErrorCode.UNAUTHENTICATED)
        }

        is PreAuthorizeSpELParser.EvaluationResult.Denied.Unauthorized -> {
            log.warn("인가 실패: {}", principal)
            throw PreAuthorizeErrorException(PreAuthorizeErrorCode.FORBIDDEN)
        }
    }

    private companion object {
        inline fun <reified T : Annotation> Method.requireAnnotation(): T =
            getAnnotation(T::class.java)
                ?: throw IllegalStateException("Required annotation ${T::class.simpleName} not found")

        fun Array<Any>.findPrincipal(): Principal? = asSequence()
            .filterIsInstance<Principal>()
            .firstOrNull()
    }
}